File: auxen/matching.py
"""Track matching and source-priority logic for the Auxen music player."""

from __future__ import annotations

import re

from auxen.models import Source, SourcePriority, Track

# Pre-compiled patterns for normalization.
_FEAT_PATTERN = re.compile(r"\b(?:feat\.?|featuring)\b", re.IGNORECASE)
_NON_ALNUM_PATTERN = re.compile(r"[^a-z0-9 ]")
_MULTI_SPACE_PATTERN = re.compile(r" {2,}")


def normalize_for_matching(text: str) -> str:
    """Normalize a string for fuzzy track matching.

    Steps:
        1. Strip leading/trailing whitespace and lowercase.
        2. Normalize "feat." / "featuring" variations to "ft".
        3. Remove non-alphanumeric characters (except spaces).
        4. Collapse multiple spaces to a single space.
    """
    result = text.strip().lower()
    result = _FEAT_PATTERN.sub("ft", result)
    result = _NON_ALNUM_PATTERN.sub(" ", result)
    result = _MULTI_SPACE_PATTERN.sub(" ", result)
    return result.strip()


def tracks_match(a: Track, b: Track, threshold: int = 85) -> bool:
    """Return True when two tracks are considered the same song.

    First checks for an exact match after normalization.  If that fails,
    uses fuzzy comparison via ``thefuzz.fuzz.ratio`` where *both* title
    and artist must score >= *threshold*.

    Gracefully returns ``False`` if ``thefuzz`` is not installed.
    """
    norm_title_a = normalize_for_matching(a.title)
    norm_title_b = normalize_for_matching(b.title)
    norm_artist_a = normalize_for_matching(a.artist)
    norm_artist_b = normalize_for_matching(b.artist)

    # Fast path: exact match after normalization.
    if norm_title_a == norm_title_b and norm_artist_a == norm_artist_b:
        return True

    # Fuzzy path.
    try:
        from thefuzz import fuzz  # noqa: WPS433
    except ImportError:
        return False

    title_score = fuzz.ratio(norm_title_a, norm_title_b)
    artist_score = fuzz.ratio(norm_artist_a, norm_artist_b)
    return title_score >= threshold and artist_score >= threshold


def pick_preferred_track(
    track_a: Track,
    track_b: Track,
    priority: SourcePriority,
) -> Track:
    """Return the preferred track according to *priority*.

    * ``PREFER_LOCAL``  -- return whichever track is local.
    * ``PREFER_TIDAL``  -- return whichever track is from Tidal.
    * ``PREFER_QUALITY`` -- return the track with the higher
      ``quality_score``.  On a tie, return *track_a*.
    * ``ALWAYS_ASK``    -- return the local track as a default (the
      caller is responsible for showing a prompt to the user).
    """
    if priority == SourcePriority.PREFER_LOCAL:
        if track_a.is_local:
            return track_a
        return track_b

    if priority == SourcePriority.PREFER_TIDAL:
        if track_a.is_tidal:
            return track_a
        return track_b

    if priority == SourcePriority.PREFER_QUALITY:
        if track_b.quality_score > track_a.quality_score:
            return track_b
        return track_a

    # ALWAYS_ASK: return the local track as default.
    if track_a.is_local:
        return track_a
    return track_b

File: android/app/src/main/kotlin/io/github/auxen/matching/Matching.kt
package io.github.auxen.matching

import io.github.auxen.model.SourcePriority
import io.github.auxen.model.Track

/**
 * Track matching and source-priority logic — Kotlin port of
 * `auxen/matching.py` from the desktop app.
 */

private val FEAT_PATTERN = Regex("""\b(?:feat\.?|featuring)\b""", RegexOption.IGNORE_CASE)
private val NON_ALNUM_PATTERN = Regex("[^a-z0-9 ]")
private val MULTI_SPACE_PATTERN = Regex(" {2,}")

/**
 * Normalize a string for fuzzy track matching: trim + lowercase, fold
 * "feat."/"featuring" to "ft", strip non-alphanumerics, collapse spaces.
 */
fun normalizeForMatching(text: String): String {
    var result = text.trim().lowercase()
    result = FEAT_PATTERN.replace(result, "ft")
    result = NON_ALNUM_PATTERN.replace(result, " ")
    result = MULTI_SPACE_PATTERN.replace(result, " ")
    return result.trim()
}

/**
 * Similarity score in 0..100, matching `thefuzz.fuzz.ratio` (the indel /
 * normalized-Levenshtein ratio): round(200 * LCS / (|a| + |b|)), where
 * "round" is Python's `round()` — round-half-to-even (banker's rounding),
 * not round-half-away-from-zero. E.g. a raw ratio of 12.5 rounds to 12,
 * not 13.
 */
fun fuzzRatio(a: String, b: String): Int {
    val lensum = a.length + b.length
    if (lensum == 0) return 100
    return Math.rint(200.0 * lcsLength(a, b) / lensum).toInt()
}

/** Longest-common-subsequence length, O(min) two-row DP. */
private fun lcsLength(a: String, b: String): Int {
    if (a.isEmpty() || b.isEmpty()) return 0
    var prev = IntArray(b.length + 1)
    var curr = IntArray(b.length + 1)
    for (i in a.indices) {
        for (j in b.indices) {
            curr[j + 1] = if (a[i] == b[j]) prev[j] + 1 else maxOf(prev[j + 1], curr[j])
        }
        val tmp = prev
        prev = curr
        curr = tmp
    }
    return prev[b.length]
}

/**
 * True when two tracks are considered the same song: exact match after
 * normalization, else both title and artist fuzzy scores >= [threshold].
 */
fun tracksMatch(a: Track, b: Track, threshold: Int = 85): Boolean {
    val titleA = normalizeForMatching(a.title)
    val titleB = normalizeForMatching(b.title)
    val artistA = normalizeForMatching(a.artist)
    val artistB = normalizeForMatching(b.artist)

    if (titleA == titleB && artistA == artistB) return true

    return fuzzRatio(titleA, titleB) >= threshold && fuzzRatio(artistA, artistB) >= threshold
}

/**
 * Return the preferred of two duplicate tracks according to [priority].
 * ALWAYS_ASK returns the local track as a default; the caller is
 * responsible for prompting the user.
 */
fun pickPreferredTrack(trackA: Track, trackB: Track, priority: SourcePriority): Track = when (priority) {
    SourcePriority.PREFER_LOCAL -> if (trackA.isLocal) trackA else trackB
    SourcePriority.PREFER_TIDAL -> if (trackA.isTidal) trackA else trackB
    SourcePriority.PREFER_QUALITY -> if (trackB.qualityScore > trackA.qualityScore) trackB else trackA
    SourcePriority.ALWAYS_ASK -> if (trackA.isLocal) trackA else trackB
}
